server/modules/routing/dbseerroute/dbseerroute.c

```diff
@@ -0,0 +1,1181 @@
+ * @param       message         The error message to reply
+ * @param       problem_dcb     The DCB related to the error
+ * @param       action     	The action: ERRACT_NEW_CONNECTION or ERRACT_REPLY_CLIENT
+ * @param	succp		Result of action: true if router can continue
+ *
+ */
+static void handleError(ROUTER *instance, void *router_session, GWBUF *errbuf,
+                        DCB *problem_dcb, error_action_t action, bool *succp)
+
+{
+    DCB *client_dcb;
+    SESSION *session = problem_dcb->session;
+    session_state_t sesstate;
+    ROUTER_CLIENT_SES *router_cli_ses = (ROUTER_CLIENT_SES *) router_session;
+
+    /** Don't handle same error twice on same DCB */
+    if (problem_dcb->dcb_errhandle_called)
+    {
+        /** we optimistically assume that previous call succeed */
+        *succp = true;
+        return;
+    }
+    else
+    {
+        problem_dcb->dcb_errhandle_called = true;
+    }
+    spinlock_acquire(&session->ses_lock);
+    sesstate = session->state;
+    client_dcb = session->client_dcb;
+
+    if (sesstate == SESSION_STATE_ROUTER_READY)
+    {
+        CHK_DCB(client_dcb);
+        spinlock_release(&session->ses_lock);
+        client_dcb->func.write(client_dcb, gwbuf_clone(errbuf));
+    }
+    else
+    {
+        spinlock_release(&session->ses_lock);
+    }
+
+    if (DCB_ROLE_CLIENT_HANDLER == problem_dcb->dcb_role)
+    {
+        dcb_close(problem_dcb);
+    }
+    else if (router_cli_ses && problem_dcb == router_cli_ses->backend_dcb)
+    {
+        router_cli_ses->backend_dcb = NULL;
+        dcb_close(problem_dcb);
+    }
+
+    /** false because connection is not available anymore */
+    *succp = false;
+}
+
+/** to be inline'd */
+
+/**
+ * @node Acquires lock to router client session if it is not closed.
+ *
+ * Parameters:
+ * @param rses - in, use
+ *
+ *
+ * @return true if router session was not closed. If return value is true
+ * it means that router is locked, and must be unlocked later. False, if
+ * router was closed before lock was acquired.
+ *
+ *
+ * @details (write detailed description here)
+ *
+ */
+static bool rses_begin_locked_router_action(ROUTER_CLIENT_SES* rses)
+{
+    bool succp = false;
+
+    CHK_CLIENT_RSES(rses);
+
+    if (rses->rses_closed)
+    {
+        goto return_succp;
+    }
+    spinlock_acquire(&rses->rses_lock);
+    if (rses->rses_closed)
+    {
+        spinlock_release(&rses->rses_lock);
+        goto return_succp;
+    }
+    succp = true;
+
+return_succp:
+    return succp;
+}
+
+/** to be inline'd */
+
+/**
+ * @node Releases router client session lock.
+ *
+ * Parameters:
+ * @param rses - <usage>
+ *          <description>
+ *
+ * @return void
+ *
+ *
+ * @details (write detailed description here)
+ *
+ */
+static void rses_end_locked_router_action(ROUTER_CLIENT_SES* rses)
+{
+    CHK_CLIENT_RSES(rses);
+    spinlock_release(&rses->rses_lock);
+}
+
+static uint64_t getCapabilities(void)
+{
+    return RCAP_TYPE_NONE;
```

Comment:
This should return the statement tracking capability defined in `routing.h` if complete SQL statements are needed.